### ISSUE TITLE
aws_acm module

### ISSUE THREAD

**Issue by mdavis-xyz** [ansible/ansible#60120]
##### SUMMARY

The `aws_acm_facts` module is used to fetch data about certs currently in ACM.

I would love it if there was a module for uploading certs to ACM.

Currently my workaround is to use a `shell` task to invoke the aws cli. But that uploads duplicates.
I think a `aws_acm` module should first check what certs exist. If the cert we want to upload already has been uploaded, nothing should happen. And if the cert we want to upload is different to one which is already uploaded, under the same domain name, we should delete the existing one and upload the new one. But some people may have other desired behavior for this case. Let me know.

I'm considering writing this myself. Is there anything I need to watch out for? Can I just copy and paste the `aws_acm_facts` module and tweak that?

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

aws_acm

##### ADDITIONAL INFORMATION

I use an Ansible playbook to build and upload Lambda functions, deploy a CloudFormation stack etc. I want to provide a cert for use by an ELB listener. To do that I need to upload a cert and pass the ARN of that cert to the Ansible task which invoked CloudFormation.

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/cloud/amazon/aws_acm_info.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/aws_acm_info.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mlda065, just so you are aware we have a dedicated Working Group for aws.
You can find other people interested in this in `#ansible-aws` on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mlda065** [comment]
I want this module to first fetch the data about what certs exist in ACM. Is there some way my module can import and use the `.py` file for `aws_acm_facts`?

**Comment by willthames** [comment]
@mida065 not in that way - you could refactor both modules to use a shared `module_utils/aws/acm.py` file though?

**Comment by mlda065** [comment]
Ah, like [this](https://github.com/ansible/ansible/blob/05c6ff79f9860dbd6b43cb4914ee749baf65b9f7/lib/ansible/modules/cloud/amazon/cloudfront_invalidation.py#L137)?

Ok. I'll give this a go.

**Comment by willthames** [comment]
Yep, that's the kind of thing - just move the methods from aws_acm_facts and then reuse in both

**Comment by mlda065** [comment]
I was just reading [this developer guide](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html).

> If you are asking ‘how can I have a module execute other modules’ … you want to write a role.

Is that applicable? I don't know what a role is, in the context of a standard building block. (I've used 'roles' to split up a huge playbook into groups of tasks.)

**Comment by willthames** [comment]
@mlda065 - I think that advice is more if you're looking to build a module that is effectively equivalent to executing more than one module.

**Comment by mlda065** [comment]
I'm looking at the tests for AWS modules. Am I correct in thinking that `aws_acm_facts` has no integration test?

**Comment by willthames** [comment]
@mlda065 - looks like I got it into 2.3, before we started mandating tests for all AWS feature requests after so many bugs in 2.4!

**Comment by mlda065** [comment]
Ah, so I'll have to write one from scratch? (Probably copying another AWS test.)

**Comment by willthames** [comment]
I am afraid so. But you'll need to test that your new module works one way or another, and the test suites are a really good starting point for doing that anyway

**Comment by mlda065** [comment]
Ok, I've almost finished writing the module, and a manually invoked integration test.

But I'm running into a bug with my code. How can I debug what my module is doing?

I put `module.debug('blah')` statements throughout, and invoke with `ansible-playbook -vvv`, but I don't see anything printed. 

I just want to have a few print statements. How do I do that? (Even if I have to delete them before merging.)

**Comment by willthames** [comment]
@mlda065 try running with `ANSIBLE_DEBUG=1`

Otherwise try the Troubleshooting section of http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html

**Comment by mlda065** [comment]
Thanks, that article is helpful, but I'm still having trouble. I think the article is out of date.

`test/integration/cloud-config-aws.yml.template` has been changed to `test/integration/cloud-config-aws.ini.template`. Consequently the `ansible –m shell ...` command fails, even when changing the file path argument. So I can't test that my permissions are set up right.

It’s not clear whether I need to update just the `{{ }}` variables, or also the `@` ones, inside that template.

When I run the example `ec2_group` test, Ansible skips all tasks. The error message suggests adding `--allow-unstable`. But when I do that I get dozens of error messages, and the test fails.

There are *a lot* of error messages, mostly about missing arguments. But they're ignored.
I think the key one is:

```
TASK [ec2_group : determine if there is a default VPC] *************************
fatal: [testhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'aws_account_attribute'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Failed to describe account attributes: An error occurred (AuthFailure) when calling the DescribeAccountAttributes operation: AWS was not able to validate the provided access credentials"}
```

That varies depending on what I put in the .ini file.

**What should my .ini file look like?**
  * should I change the `[default]` bit to match my `~/.aws/credentials` profile name?
  * Should I change the `@` variables?
  * What do I put as the "security_token"?
  * Should I change the `ec2_... : {{ ... }}` parameters?

Here is some other feedback about your article. 

* `--group_name` should be `--group-name`
* For “writing your own tests”, the directory is not `/test/integration/module_name` it’s `/test/integration/targets/module_name`

**Comment by willthames** [comment]
oh, thanks for the feedback, it is a little out of date but it was about how to do the troubleshooting rather than get AWS setup (but I should fix it so that it's accurate for both)

If you've followed the steps for creating the ansible test user, then you just need to replace the `@` variables - the `{{ }}` variables will just reference the values from the `@` variables

```
[default]
aws_access_key: AKIA...
aws_secret_key: some_secret_value
security_token: ""
aws_region: us-east-2

ec2_access_key: {{ aws_access_key }}
ec2_secret_key: {{ aws_secret_key }}
ec2_region: {{ aws_region }}
```

should all work

**Comment by mlda065** [comment]
I have written one playbook to test both `aws_acm` and `aws_acm_info`. What do I need to do to the `alias` files in `test/integration/targets` to tell the framework that the test for one is the same as the other?

**Comment by willthames** [comment]
@mlda065 I think you just put the name of the other one in the aliases file

**Comment by mlda065** [comment]
My playbook passes when run manually. But when I run it as an integration test with docker it fails.

I've create a PR so you can have a look.
Can you please help me figure out why my integration test fails when hooked in to the framework, but passes when run as a standalone playbook? (Details in the PR)

**Comment by Xyrion** [comment]
Any update on this?

**Comment by matt-telstra** [comment]
I (same person as mlda065) have written [a pull request](https://github.com/ansible/ansible/pull/60552).

The code works. The tests all pass. But the people validating my work are hitting undocumented AWS limits. I'm not sure that the status is on that.

If you want to use the module today, you can pull the code in the pull request and run Ansible from that commit.

**Comment by matt-telstra** [comment]
The pull request has now been merged into the devel branch.

**Comment by Xyrion** [comment]
That's great news!!
